usersync/cookie.go

```diff
@@ -20,70 +19,140 @@ const uidTTL = 14 * 24 * time.Hour
+// To write an instance onto a response, use WriteCookie.
 type Cookie struct {
-	uids   map[string]uidWithExpiry
+	uids   map[string]UIDEntry
 	optOut bool
 }
 
-// uidWithExpiry bundles the UID with an Expiration date.
-// After the expiration, the UID is no longer valid.
-type uidWithExpiry struct {
+// UIDEntry bundles the UID with an Expiration date.
+type UIDEntry struct {
 	// UID is the ID given to a user by a particular bidder
 	UID string `json:"uid"`
 	// Expires is the time at which this UID should no longer apply.
 	Expires time.Time `json:"expires"`
 }
 
-// ParseCookieFromRequest parses the UserSyncMap from an HTTP Request.
-func ParseCookieFromRequest(r *http.Request, cookie *config.HostCookie) *Cookie {
-	if cookie.OptOutCookie.Name != "" {
-		optOutCookie, err1 := r.Cookie(cookie.OptOutCookie.Name)
-		if err1 == nil && optOutCookie.Value == cookie.OptOutCookie.Value {
-			pc := NewCookie()
-			pc.SetOptOut(true)
-			return pc
-		}
-	}
-	var parsed *Cookie
-	uidCookie, err2 := r.Cookie(uidCookieName)
-	if err2 == nil {
-		parsed = ParseCookie(uidCookie)
-	} else {
-		parsed = NewCookie()
-	}
-	// Fixes #582
-	if uid, _, _ := parsed.GetUID(cookie.Family); uid == "" && cookie.CookieName != "" {
-		if hostCookie, err := r.Cookie(cookie.CookieName); err == nil {
-			parsed.TrySync(cookie.Family, hostCookie.Value)
-		}
+// NewCookie returns a new empty cookie.
+func NewCookie() *Cookie {
+	return &Cookie{
+		uids: make(map[string]UIDEntry),
 	}
-	return parsed
 }
 
-// ParseCookie parses the UserSync cookie from a raw HTTP cookie.
-func ParseCookie(httpCookie *http.Cookie) *Cookie {
-	jsonValue, err := base64.URLEncoding.DecodeString(httpCookie.Value)
+// ReadCookie reads the cookie from the request
+func ReadCookie(r *http.Request, decoder Decoder, host *config.HostCookie) *Cookie {
+	if hostOptOutCookie := checkHostCookieOptOut(r, host); hostOptOutCookie != nil {
+		return hostOptOutCookie
+	}
+
+	// Read cookie from request
+	cookieFromRequest, err := r.Cookie(uidCookieName)
 	if err != nil {
-		// corrupted cookie; we should reset
 		return NewCookie()
 	}
+	decodedCookie := decoder.Decode(cookieFromRequest.Value)
 
-	var cookie Cookie
-	if err = json.Unmarshal(jsonValue, &cookie); err != nil {
-		// corrupted cookie; we should reset
-		return NewCookie()
+	return decodedCookie
+}
+
+// PrepareCookieForWrite ejects UIDs as long as the cookie is too full
+func (cookie *Cookie) PrepareCookieForWrite(cfg *config.HostCookie, ttl time.Duration, encoder Encoder) string {
+	encodedCookie := encoder.Encode(cookie)
+	isCookieTooBig := len(encodedCookie) > cfg.MaxCookieSizeBytes && cfg.MaxCookieSizeBytes > 0
+	uuidKeys := sortUIDs(cookie.uids, isCookieTooBig)
+
+	i := 0
+	for isCookieTooBig && len(cookie.uids) > 0 {
+		uidToDelete := uuidKeys[i]
+		delete(cookie.uids, uidToDelete)
+
+		encodedCookie = encoder.Encode(cookie)
+		isCookieTooBig = len(encodedCookie) > cfg.MaxCookieSizeBytes && cfg.MaxCookieSizeBytes > 0
```

Comment:
At this point we know `cfg.MaxCookieSizeBytes > 0` evaluates to true. We can probably ommit this check to optimize a bit and and make this line a bit less verbose
```
71         encodedCookie = encoder.Encode(cookie)
72 -       isCookieTooBig = len(encodedCookie) > cfg.MaxCookieSizeBytes && cfg.MaxCookieSizeBytes > 0
   +       isCookieTooBig = len(encodedCookie) > cfg.MaxCookieSizeBytes
73         i++
usersync/cookie.go
```